### ISSUE TITLE
Chart annotations not visible in previews

### ISSUE THREAD

**Issue by SilviuDinu** [Codeinwp/visualizer#600]
### Testing environment
https://fascinating-racoon.jurassic.ninja 
WP user: demo WP/SSH password: SCLJjdnAzihN

### Issue
I added annotations for a column chart and Saved it. The annotations weren't visible in the chart's preview in the Chart Library or in the Gutenberg block., but the chart is correctly displayed on the frontend.

The annotations appeared in the preview only after adding another chart (a new one) that had Annotations as well.

Videos:

- https://share.vidyard.com/watch/n3BvWNKR136qoQJ4MaerTm (Chart Library)

- https://share.vidyard.com/watch/SCMkbhqrTrdditfHELq1h5 (Chart Library)

- https://share.vidyard.com/watch/t1S8vHmFmcjufnGbQRpt5v (Gutenberg block)

**Comment by contactashish13** [comment]
@SilviuDinu I've fixed this and merged it into the 3.4.0 branch.

It will be a good idea to test a library with
- only google charts (charts without annotations) 
- only google charts (charts only with annotations) 
- only google charts (charts with and without annotations) 
- only chartjs charts
- only table charts
- a mix of the above

**Comment by DanyCaissy** [comment]
@contactashish13 

If I clone a chart with annotations multiple times then hit F5 then the library page will load indefinitely. This didn't happen when I added new chart afterwards, but I hit F5 and then it never worked again. It seems as if the F5 itself broke it.

https://literary-shrimp.jurassic.ninja 
WP user: demo WP/SSH password: LaYOyDnv8XNP

https://literary-shrimp.jurassic.ninja/wp-admin/admin.php?page=visualizer

**Comment by DanyCaissy** [comment]
Just in case, this is using chrome, and it looks like that: https://i.imgur.com/8mVZhV3.png

They never load

**Comment by contactashish13** [comment]
@DanyCaissy this would need a video to see the process you followed. I cannot replicate it.

**Comment by DanyCaissy** [comment]
@contactashish13 

I tried to test this, but it seems charts just don't load at all in the current version of 3.4.

https://fair-antelope.jurassic.ninja
WP user: demo WP/SSH password: VDKSpW7Hkptv

To reproduce:

I just created a new instance, added 3.4, and created a default piechart. After F5 chart doesn't load in library in either chrome or firefox

**Comment by contactashish13** [comment]
@DanyCaissy I cannot reproduce this. I created a line chart and then refreshed. You can log in as well and click refresh. This cannot be understood or reproduced without a video.

**Comment by DanyCaissy** [comment]
@contactashish13 

Well, they just don't load for me, in all 3 browsers that I've tried:

![2019-12-12_21-53-40](https://user-images.githubusercontent.com/5183243/70717730-eb7e2980-1d29-11ea-951d-df91e6e74476.gif)

I've even tried to use a VPN and changed location.

@SilviuDinu Can you login to that instance and see if you can see the charts in the library after hitting F5?

**Comment by SilviuDinu** [comment]
@DanyCaissy @contactashish13 

I tested the fair-antelope instance and I've created a video with the results. 

Video link -> https://share.vidyard.com/watch/DoEsKUxjfwdyEHT1YpqRLr?

I have to press F5 a few times before the issue appears. 
On the first browser tab, I had to press F5 twice in 1 second for the issue to occur. 
However, on the incognito window, I pressed F5 normally (like any other user would do) and the issue still occurred. No error in the console.

**Comment by SilviuDinu** [comment]
> @SilviuDinu I've fixed this and merged it into the 3.4.0 branch.
> 
> It will be a good idea to test a library with
> 
> * only google charts (charts without annotations)
> * only google charts (charts only with annotations)
> * only google charts (charts with and without annotations)
> * only chartjs charts
> * only table charts
> * a mix of the above

The issue with endlessly loading the preview happened to me with every scenario from above. I had to create a new instance for each scenario because even after deleting the plugin and install it again. the newly created charts had this issue from the start.

I did not record new videos because is the same as in my last comment, only with multiple chart types. Here are some of my instances where I tested (cc @DanyCaissy if you could check)

both annotations: https://ashamed-canidae.jurassic.ninja/wp-admin/admin.php?page=visualizer
user: demo
pass: 7H4q4Umn99LV

one with annotations, the other without: https://better-koi.jurassic.ninja/wp-admin/admin.php?page=visualizer
user: demo
pass: tkPRMnA8TUZS

2x chartJS: https://dangerous-giraffe.jurassic.ninja/wp-admin/admin.php?page=visualizer
user: demo
pass: OGKVOBrQa0Hn

2x different table charts: https://unsightly-marmot.jurassic.ninja/wp-admin/admin.php?page=visualizer
user: demo
pass: igevTn6LV5rD

**Comment by contactashish13** [comment]
@SilviuDinu @DanyCaissy Let me explain why this happens so that when you try to reproduce you can tell me the **exact sequence of events** because I cannot reproduce this. When charts are loaded in the library, each chart is called N times where N = the number of libraries being loaded in the page. That is why I've suggested testing with single and multiple libraries. When a chart is loaded, a local data storage is created in the browser that keeps a flag of whether the initialization sequence has been called or not. If it has been called before, it is aborted. The flag (`viz-facade-loaded`) is reset 2 seconds after the page loads so any use of F5 within 2 seconds will result in the charts not loading again. That is why instead of reinstalling the plugin it is enough to clear the HTML5 local data storage of the browser. 

![image](https://user-images.githubusercontent.com/12953439/70799740-7e719f00-1dd0-11ea-80ef-232d59f369ee.png)

That is why it is important to understand how to reproduce this issue along with the time of each sequence of actions. F5 within 2 seconds will make this happen, F5 after 2 seconds should not.

**Comment by DanyCaissy** [comment]
It used to only happen "before 2 seconds" for me, kinda, now it just happens no matter what, as I explained above.

Here is a gif of the how to reproduce it, from the very start, the creation of the instance even:

![2019-12-16_23-51-18](https://user-images.githubusercontent.com/5183243/70921606-96ab1d80-205f-11ea-8f59-fcc6720e4d3a.gif)


https://marked-stork.jurassic.ninja
WP user: demo WP/SSH password: LkjTX4yhEZz4

This was in chrome but happens in other browsers (firefox) or incognito windows and also if I use a vpn. 

We can look at the edge cases you mentioned afterwards, but firstly this should be resolved as it happens on the base case already.

**Comment by DanyCaissy** [comment]
@contactashish13 

Gif didn't upload the first time, here it is:

![2019-12-16_23-51-18](https://user-images.githubusercontent.com/5183243/70921712-ce19ca00-205f-11ea-841f-fab8db39be74.gif)

**Comment by contactashish13** [comment]
@DanyCaissy I tried this on a fresh instance and cannot reproduce this. I have a feeling you are refreshing this within 2 seconds again. Can you please open the dev console to the tab in my screenshot and see how that local storage behaves?

**Comment by DanyCaissy** [comment]
@contactashish13 

Why would you say I'm refreshing within 2 seconds when the previous video shows I'm waiting longer?

Video in the zip:

[2019-12-21_16-59-25.zip](https://github.com/Codeinwp/visualizer/files/3991237/2019-12-21_16-59-25.zip)

I'm using both chrome and then firefox, waiting way longer than two seconds, the local storage is open also.

The charts just wont display at all.

**Comment by contactashish13** [comment]
@DanyCaissy there was no timeline in the GIF you had attached so I had no way of knowing how long you are waiting. In the video you have attached now, the timeline helps to understand this better.

I dont know why your instance on your computer behaves differently from the same instance on mine. Video attached.
[video.zip](https://github.com/Codeinwp/visualizer/files/3992166/video.zip)

**Comment by DanyCaissy** [comment]
@contactashish13 

This was not on my computer, this was on https://marked-stork.jurassic.ninja

**Comment by contactashish13** [comment]
@DanyCaissy please view the video I have attached as well. I have mentioned **instance** on different computeres

**Comment by DanyCaissy** [comment]
@contactashish13 

I don't know either but if it this happens to me, it can happen to other users, so it should be fixed regardless. This happens to me on multiple browsers, IPs (different countries + vpn), and multiple instances and can be reproduced when I follow the steps.

If you think this is an issue with my laptop, maybe we can ask least 5 people to reproduce these steps, if none of them encounter this issue, we can consider releasing.

**Comment by contactashish13** [comment]
@SilviuDinu @selul can you please try the below credentials and see if you are able to view the chart when you press F5? @Codeinwp/support @mileamadalin can you help as well?

> https://marked-stork.jurassic.ninja
> WP user: demo WP/SSH password: LkjTX4yhEZz4

**Comment by DanyCaissy** [comment]
Also, try different browsers, different interval (F5 after 1s, F5 after 5s, etc.) and note if there's any differences please.

**Comment by SilviuDinu** [comment]
I can't replicate it anymore unless I'm pressing F5 within less than ~2 secs (on all browsers from below). The chart is visible to me otherwise (if I wait more than 2 secs between pressing F5).

chrome -> https://share.vidyard.com/watch/KoEKZ5dGw2GLRRPBX8nCoZ?
edge -> https://share.vidyard.com/watch/ZYUeQ9dxxGb9f1UykMwtiy?
firefox -> https://share.vidyard.com/watch/jn6X86yDwBK1PSAXGUDujg?

**Comment by DanyCaissy** [comment]
@selul  @Codeinwp/support @mileamadalin 

We really need more users to try the steps above, please attempt the steps above and report back.

**Comment by DanyCaissy** [comment]
@contactashish13 Note that putting the current public visualizer version on the same instance will fix the charts and they will load no matter what, even if is I hit F5 before 2 seconds. On the new version, F5 before or after 2 seconds will not load no matter what.

Before release, we need to ensure that this version works as well as the previous (meaning charts should load, even if F5 is hit within less than 2s).

**Comment by mileamadalin** [comment]
I will try the steps above in the upcoming days.

**Comment by contactashish13** [comment]
@DanyCaissy refresh before 2s will not work  and there is no reason to expect it to, either, considering the change in the code and the reason it has been put in place as explained in https://github.com/Codeinwp/visualizer/issues/600#issuecomment-565421196. 

However the charts will load when the next time the user does not refresh within 2s and allows the charts to load. Refresh within 2s is not a common use case anyway.

**Comment by DanyCaissy** [comment]
@mileamadalin @Codeinwp/support 

Could you please test this (and ask another member in the support team also) ?

We cannot release until this has been verified by multiple users.

**Comment by DanyCaissy** [comment]
Just want to note again, if I clear my cache and refresh only after 2 seconds each time, my library is fine.

The issue is that, if I do it within 2 seconds ONCE, the entire instance will never refresh properly again, before or after 2s, until I refresh the cache. It's like refreshing within 2s 'corrupts' the cache until it is cleared again.

**Comment by rodica-andronache** [comment]
@contactashish13 @DanyCaissy I've tested this and here is what I saw:

- for the annotations issue. I've tried creating a new **Column** type chart, go to **Series Settings** and tried to update the **color** of one of the elements and add **annotations**. An error appeared in the console, which I think is the cause for the color and annotation not to apply in the preview.
Here's a video of what I've tested, where you can see the error in the console https://www.loom.com/share/0a7f8a88ec3b4e7b90994f64fcef659f

- for the charts not loading. From what I've seen, I think this is not necessarily related to the annotations. What I've done is:
1. create a chart ( no matter its type )
2. duplicate it -> everything is ok
3. duplicate it again -> all the charts stop working.
At this point, no matter what I do, if I add a new chart, delete those charts, refresh the page, the charts don't load, unless I clear the local data storage.

Here's a video of that https://www.loom.com/share/32803286bb03404ea0a7238f3a31b2f6

Let me know if that helps or if there's something else I can do

**Comment by contactashish13** [comment]
@rodica-andronache Can you please confirm that the error in your first video is not a consequence of https://github.com/Codeinwp/visualizer/issues/602#issuecomment-571579735?

To reproduce your second video I need the page to load super quick (within 1s) and my net connection conspires against it so I'm unable to replicate your steps even on a clean jurassic instance. However, I have pushed in a commit that offers a workaround. Even if the library does not load once, a new refresh will fix it. Let me know if PR https://github.com/Codeinwp/visualizer/pull/606 does the trick. 

PS - I see that you are using the Gutenberg editor in the video. It would be great to stick to only the classic editor for any issues as I have full knowledge of the code there compared to the Gutenberg part :) But you can obviously open a new issue for those so that they are not lost - maybe prefix it with "Gutenberg"?

**Comment by rodica-andronache** [comment]
@contactashish13 

1.  Indeed, the first error is happening only when you have `define('SCRIPT_DEBUG', true);` set and is working ok with the fix here https://github.com/Codeinwp/visualizer/pull/605

2. But without `define('SCRIPT_DEBUG', true);` I still think the annotations don't work properly, at least, from what I understand, with this PR https://github.com/Codeinwp/visualizer/pull/561
You can see in the video, that the annotations I set for **Cats** just make the corresponding line disappear, the annotations I set for **Blanket1** apply to the line corresponding to **Cats** and **Blanket2** makes the corresponding line disappear. Let me know if I'm not testing properly, or if that's normal.
https://www.loom.com/share/5530c3795bc6460396a9b59bcad40efc

2. For the issue with the charts not loading, https://github.com/Codeinwp/visualizer/pull/606 seems to fix this. Indeed had few cases where the issue occurred while creating/duplicating or deleting charts to fast, but I can't say an exact pattern, and it doesn't look to be a real issue from my side.

PS. When you mean the Gutenberg editor, you mean the WP editor, or using the Gutenberg plugin? I should be testing with the plugin for the Classic editor installed?

**Comment by contactashish13** [comment]
@rodica-andronache 

1 and 3 - Perfect!

2. I've tried to explain how annotations work here: https://github.com/Codeinwp/visualizer/issues/214#issuecomment-552338251 Let me know if you need more clarity and I can explain it on slack.

PS - Sorry, I meant that when you try to insert the chart to test it on the front end, it would be preferable to use the classic editor so that any idiosyncrasy introduced by our Gutenberg code does not skew your tests. Of course, you can also try using the block editor and if you spot any issues they can be added to the issue list.

**Comment by rodica-andronache** [comment]
@contactashish13 thanks, I've understood now how annotations work and everything looks good from what I've tested
